dbms/src/DataStreams/MultiplexInputStream.h

```diff
@@ -33,43 +33,43 @@ class MultiPartitionStreamPool
 public:
     MultiPartitionStreamPool() = default;
 
+    void cancel(bool kill)
+    {
+        bool old_val = false;
+        if (!is_cancelled.compare_exchange_strong(old_val, true, std::memory_order_seq_cst, std::memory_order_relaxed))
+            return;
+
+        for (auto & stream : added_streams)
```

Comment:
Seems `std::unique_lock lk(mu);` is needed here.
Otherwise there is a data race with `Stream::read::pickOne`